src/core/kd_tree.rs

```diff
@@ -24,88 +24,123 @@ enum Node<T: Geometry> {
+        IntersectionIterator {
+            node_stack,
+            items: vec![],
+            ray,
+        }
+    }
+
+    fn process_items(&mut self) -> Option<Intersection> {
+        while self.items.len() > 0 {
+            let item = self.items.pop().unwrap();
+            match item.intersect(&self.ray) {
+                Some(intersection) => {
+                    self.ray.t_max = intersection.distance;
+                    return Some(intersection);
+                }
+                None => {}
+            }
+        };
+        None
+    }
+}
+
+impl<'a, T> Iterator for IntersectionIterator<'a, T>
+    where T: Geometry {
+    type Item = Intersection;
+
+    // pbrt pg. 240
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut hit: Option<Self::Item> = self.process_items();
+        while hit.is_none() && self.node_stack.len() > 0 {
+            let (node, t_min, t_max) = self.node_stack.pop().unwrap();
+            if t_min < self.ray.t_max {
+                match node {
+                    &Node::Internal(axis, distance, ref left, ref right) => {
+                        let axis_index = axis as usize;
+                        let origin_component = self.ray.origin[axis_index];
+                        let direction_component = self.ray.direction[axis_index];
+                        let (near, far) = if (origin_component < distance) || (origin_component == distance && direction_component <= 0f64) {
+                            (left, right)
+                        } else {
+                            (right, left)
+                        };
+
+                        let t_plane = (distance - origin_component) / direction_component;
+                        if t_plane > t_max || t_plane <= 0f64 {
+                            // t_plane > t_max means we hit the plane outside the current node's bounds, so skip far.
+                            // t_plane <= 0 is not because the starting point of the ray is significant, but because the
+                            // sign tells us if we're pointing away from the plane and can skip far.
+                            // Note that this automatically handles both infinities.
+                            self.node_stack.push((near, t_min, t_max));
+                        } else if t_plane < t_min {
+                            // t_plane < t_min means we're poining towards the plane, but it's behind where we care about
+                            // so skip near.
+                            self.node_stack.push((far, t_min, t_max));
+                        } else {
+                            self.node_stack.push((far, t_plane, t_max));
+                            self.node_stack.push((near, t_min, t_plane));
                         }
                     }
-                },
+                    &Node::Leaf(ref items) => {
+                        self.items.extend_from_slice(items);
```

Comment:
I think is where the perf hit is.

Reply:
Yikes, this `clone`s everything in it? Very possible.